Deprecate ``qiskit.extensions``

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Deprecates the ``qiskit.extensions`` submodule. Contents are either moved to the circuit library or deprecated. In future it would be nice to build high-level synthesis passes for algorithms that synthesis the same type of operation. E.g. isometry could be a synthesis for unitaries (or for state initialization).

### Details and comments

Here's a detailed list of moves:

#### Moved to generalized gates
* ``uc<any>.py``: These are multiplexed gates, as such they are generalizations of gates we have in the standard gate library
* ``unitary.py``: General unitary gate (hence generalized gate)
* ``isometry.py``: Implements isometries, and as subset of this also unitaries, so it should live in the same place as `UnitaryGate`.
* ``mcg_up_diag.py``: Implements a multicontrolled gate up to a diagonal, hence a generalization of controlled gates. 
* ``diagonal.py``: Moved the ``DiagonalGate`` into the same file as the `Diagonal` circuit

#### Moved to data preparation
* initializer.py: Prepares an arbitray initial state and should be next to `StatePreparation`, which is a data preparation operation. Could also live next to `UnitaryGate`, but then we'd also have to move the state preparation.

#### Moved to circuit library root:
* ``hamiltonian_gate.py``: Implements an exact operator evolution and should be next to ``PauliEvolutionGate``. 

#### Deprecated:
* ``squ.py``: Implements a 1q-unitary matrix and thus a duplicate functionality of UnitaryGate. Also not used anywhere in the source code. 
* `snapshot.py`: Superseded by Aer's `save_*` instructions.

#### Monkey-patched circuit methods:
* Deprecated: ``diagonal``, ``hamiltonian``, ``uc``, ``ucrx``, ``ucry``, ``ucrz``, ``squ``, ``snapshot``
* Moved to the circuit class as they are frequently used: ``unitary``, ``initialize``

Cryoris committed:
big moves, import still works

Cryoris committed:
most tests pass (some I cannot seem to run locally)

Cryoris committed:
fix tests -- how to remove DiagonalGate?

Cryoris committed:
typehints and docs

Cryoris committed:
more type hints

Cryoris committed:
Deprecate SQU

Cryoris committed:
deprecate Snapshot

Cryoris committed:
Merge branch 'main' into deprecate-extensions

Cryoris commented on qiskit/circuit/library/__init__.py:
```diff
@@ -181,6 +181,7 @@
    :template: autosummary/class_no_inherited_members.rst
 
    Diagonal
+   DiagonalGate
```
I would've liked to deprecate this as the `Diagonal` circuit exists, but `DiagonalGate` seems to be used in the Aer simulators.

jakelishman replied:
imo it should be the circuit that gets removed not the gate, since circuits inherently define an eager synthesis, but that's not important.

Cryoris replied:
Agreed, maybe this would best be part of an overhaul to consistently use Gate or Circuits (probably the first) 🙂

ShellyGarion replied:
I agree that this is quite confusing. Maybe it's worth to add a comment somewhere about the difference between them, and which one will be deprecated?
(although we also have `Permutation` and `PermutationGate`)

alexanderivrii replied:
I think we had the plans to deprecate `Permutation` circuit once `PermutationGate` is around. Do you think it's a good idea?

Cryoris commented on qiskit/circuit/library/generalized_gates/mcg_up_to_diagonal.py:
This seems to be only used as a subroutine in the `Isometry`, this could potentially also be made private.

Cryoris commented on qiskit/circuit/library/hamiltonian_gate.py:
We could consider creating an `evolutions` submodule containing this and the `PauliEvolutionGate`.

Cryoris commented on qiskit/circuit/library/hamiltonian_gate.py:
```diff
@@ -0,0 +1,142 @@
+        data = np.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
+            raise ValueError("Input matrix is not Hermitian.")
```
Some errors were changed from `ExtensionsError` to `ValueError`, which is strictly speaking not backwards compatible. It seemed small enough to be still fine, but we could also create an intermediate class, covering both, for the deprecation period.

Cryoris commented on qiskit/circuit/library/data_preparation/initializer.py:
I was hoping most files would show up as simply moved instead of newly created (I used `git mv` to move them) because most of this content is not new.

jakelishman replied:
Looks like you put something substantial in place of the old files.  Git doesn't natively track renames, it redetects them from the tree diff.

ElePT replied:
When qiskit-nature was refactored a few releases ago, we had to make a history-preserving copy of the whole `second_q` module, and kept both the old and copied files throughout the refactoring. We managed to do it and get git to understand what we were doing (kept histories on both, new files detected as new, old files not detected as new). The approach is a bit cumbersome, it requires 2 PRs, and I am not 100% sure you would need it. But it is a way to avoid old files being detected as "new", so I thought it could make sense to share it. The approach we used is shown [here](https://github.com/qiskit-community/qiskit-nature/pull/746), and it was based on this [stackoverflow comment](https://stackoverflow.com/questions/16937359/git-copy-file-preserving-history/44036771#44036771).

Cryoris replied:
Edit: The files show up as moved now, as the original files no longer exist.

jakelishman commented on qiskit/extensions/quantum_initializer/initializer.py:
```diff
@@ -13,18 +13,14 @@
 
 
-class Initialize(Instruction):
+class Initialize(NewInitialize):
```
I don't think we need a new class here - we can just have a module-level `getattr` in all the relevant places that imports the actual class from the correct place and issues a warning.

Cryoris replied:
That could also work if we remove `import qiskit.extensions` from the `qiskit/__init__.py` and move the monkey-patching of circuit methods to the circuit library. (Otherwise any qiskit import would trigger the deprecation warnings)

jakelishman replied:
you can also at a `__getattr__` on `QuantumCircuit` that issues the deprecation warnings on the monkeypatching because yeah, we shouldn't be importing a deprecated module ourselves.

jakelishman commented on qiskit/assembler/disassemble.py:
```diff
@@ -165,16 +179,12 @@ def _experiments_to_circuits(qobj):
-            if hasattr(circuit, name):
+            # TODO remove the check that name is not in the deprecated circuit methods
+            # once the methods have been removed
+            if hasattr(circuit, name) and name not in _DEPRECATED_CIRCUIT_METHODS:
```
Strictly you might want `_DEPRECATED_CIRCUIT_METHODS` to be some sort of `set` if you want to do many lookups into it.

Cryoris committed:
Fix missing future annotations import

Cryoris committed:
minimize deprecation effort

Cryoris committed:
Merge branch 'main' into deprecate-extensions

Cryoris committed:
Change to pending deprecation, no exact-location import supported

Cryoris committed:
Merge branch 'deprecate-extensions' of github.com:Cryoris/qiskit-terra into deprecate-extensions

Cryoris committed:
Merge branch 'main' into deprecate-extensions

Cryoris committed:
fix MCG<->MGC typo and snapshot deprecation

Cryoris committed:
Merge branch 'deprecate-extensions' of github.com:Cryoris/qiskit-terra into deprecate-extensions

Cryoris committed:
fix pylint, try fixing docs

Cryoris committed:
remove gates from extensions toctree

ShellyGarion commented on qiskit/circuit/library/__init__.py:
```diff
@@ -505,8 +516,16 @@
+    UCRXGate,
+    UCRYGate,
+    UCRZGate,
 )
```
should MCGupDiag (from mcg_up_diag.py) be here too?
(I see the file, but it does not appear in the API docs)

Cryoris replied:
Since it wasn't in the API docs in qiskit-extensions I didn't include it in the library either. I think this gate is only a utility for `Isometry` and should probably be private or even a function, but this cleanup is probably better suited for a follow up 🙂

ShellyGarion commented on qiskit/extensions/quantum_initializer/squ.py:
```diff
@@ -158,6 +164,11 @@ def validate_parameter(self, parameter):
 
+@deprecate_func(
+    since="0.45.0",
+    additional_msg="Instead, append a qiskit.circuit.library.UnitaryGate to the circuit.",
```
Is it clear how to replace `squ` by `unitary`? or should some explanation be given?
Do both functions give the same output for 1-qubit unitaries?

Cryoris replied:
Replacing is straightforward, both allow the signature `Gate(unitary, qubit_index)`. SQU has two additional arguments that the unitary gate doesn't have:
* `mode` for the decomposition -- but there is only one option, so this argument is pointless 😄 (I assume the plan was to add other modes but that never happened)
* `up_to_diagonal`, which allows to implement the gate up to a diagonal. In the tests I did, however, this sometimes gave a decomposition with higher gate count than `UnitaryGate`, so I would think we can just remove it

ShellyGarion commented:
I see that you updated the legacy release notes and old release notes, but perhaps there should be new release notes about adding these gated to the circuit library?

Cryoris committed:
Merge branch 'main' into deprecate-extensions

Cryoris committed:
Add reno, fully deprecate SQU and Snapshot

alexanderivrii commented on qiskit/circuit/library/data_preparation/initializer.py:
```diff
@@ -0,0 +1,96 @@
+    """Complex amplitude initialization.
+
+    Class that initializes some flexible collection of qubit registers, implemented by calling
+    the :class:`qiskit.extensions.StatePreparation` Class.
```
Should `qiskit.extensions` here also be updated?

Cryoris replied:
Good catch, yes it should!

alexanderivrii commented on qiskit/__init__.py:
```diff
@@ -66,9 +66,6 @@
 # user config
 from qiskit import user_config as _user_config
 
-# The qiskit.extensions.x imports needs to be placed here due to the
-# mechanism for adding gates dynamically.
-import qiskit.extensions
```
This probably shows my not understanding of the deprecation process, but what happens now if someone calls `from qiskit.extensions import UnitaryGate`? Would this be still temporarily allowed?

Cryoris replied:
All imports from `qiskit.extensions` work as before. This line here was needed before, as the extensions module monkey-patched methods to the circuit, which are only available once the module is imported. 

I moved all methods directly onto the circuit, except `squ` and `snapshot`, so those work as before without needing this import. However if a user now wants to call `QuantumCircuit.squ` that doesn't work anymore... I'll add that import so it still works.

Cryoris replied:
Fixed in 3e92ab3b0bfc142ab21e8db9df8b88f7c8c57d17

alexanderivrii commented on qiskit/circuit/library/generalized_gates/uc_pauli_rot.py:
```diff
@@ -13,39 +13,36 @@
-"""
+"""Uniformly controlled Pauli rotations."""
+
+from __future__ import annotations
```
This is more ignorance on my part. Why is this needed?

Cryoris replied:
This is needed to enable type hints like `list[...]` (instead of `typing.List[...]`) or items from `collections.abc` instead of using the `typing` module in Python 3.8. For example the following code fails in Python 3.8
```python
# in 3.8 this import is needed, in 3.9 it is implied and can be removed
# from __future__ import annotations

def f(x: list[int]) -> None:
    print(x)

f([1, 2])
```

Cryoris committed:
Merge branch 'main' into deprecate-extensions

Cryoris committed:
Apply Sasha's review comments

- fix usage of .squ and .snapshot w/o import
- fix docstring usage of extensions
- fix tests

Cryoris committed:
capture snapshot deprecation warning

ElePT commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4120,6 +4122,664 @@ def pauli(
+        # Check if there is one target qubit provided
+        if not isinstance(qubit, Qubit):
+            raise QiskitError("The target qubit is not a single qubit from a QuantumRegister.")
+        return self.append(SingleQubitUnitary(unitary_matrix, mode, up_to_diagonal), [qubit], [])
```
If I am not mistaken, this line will raise a second deprecation warning as per `qiskit/extensions/quantum_initializer/squ.py L45`, I see that you catch the deprecation warning for `snapshot`. Shouldn't you have a similar filter here?

Cryoris replied:
Yeah I was a bit lazy, as I thought more deprecation warnings wouldn't be too bad 😛 I'll add a catch for the second one 🙂

ElePT replied:
well, my experience with opflow deprecation warnings tells me that they are not as innocent as they seem (they can break unit tests from other libraries that depend on qiskit), so better safe than sorry :)

Cryoris replied:
Fixed in 80ace24

ElePT commented on releasenotes/notes/deprecate-extensions-d7ab3fca19962e2e.yaml:
```diff
@@ -0,0 +1,38 @@
+  - |
+    Progress towards deprecating the :mod:`qiskit.extensions` module to streamline the structure
+    of Qiskit. The following objects have been moved to :mod:`qiskit.circuit.library`
+    and will be deprecated in :mod:`qiskit.extensions` in a future release,
```
It might be good to have a line saying explicitly that the module as a whole is pending deprecation. I know that the meaning is the same as "will be deprecated" and "progress towards deprecating" but it having the keyword could help is someone gets the `PendingDeprecationWarning` and is reading diagonally (this may or may not have been me during this review)

Cryoris replied:
Actually I think all deprecations are proper deprecations now -- so we can just write that the whole module is deprecated 👍🏻

ElePT replied:
are they? I see pending here: https://github.com/Qiskit/qiskit/pull/10725/files#diff-bf2e285d2983dfe16101ec2798b449833428a59785a1e4f70da5281dc439dec9R65-R69

ShellyGarion replied:
Perhaps it's also worth to comment on the deprecated ExtensionError? 
Since it's part of the API docs: 
https://qiskit.org/documentation/apidoc/extensions.html#qiskit.extensions.ExtensionError

Cryoris replied:
You're right, the module is pending deprecation, as the the new import path from the circuit library is just added in this release. Only some circuit methods are deprecated directly (I mixed something up with #10797 😅).

Cryoris replied:
Fixed in 80ace24

ShellyGarion commented on releasenotes/notes/deprecate-extensions-d7ab3fca19962e2e.yaml:
```diff
@@ -0,0 +1,38 @@
+    * :class:`~.library.UCRXGate`,
+    * :class:`~.library.UCRYGate`,
+    * :class:`~.library.UCRZGate`,
+    * :class:`~.library.UnitaryGate`,
```
`DiagonalGate` is missing in this list, as well as the circuit method `diagonal`.

Cryoris replied:
Added in [80ace24](https://github.com/Qiskit/qiskit/pull/10725/commits/80ace241f0ceb7026aae493c8ac7e94cbfeb058b) and e2e7807

Cryoris committed:
review comments

- capture warning of SQU
- update reno to explicitly mention pending deprecation, add DiagonalGate and ExtensionError

Cryoris committed:
missed `diagonal` method

alexanderivrii commented:
I think this is ready to be merged: all the review comments have been addressed. @Cryoris, thanks for the massive work on this, I am personally very excited to see the `qiskit.extensions` module finally being deprecated.

alexanderivrii committed:
Merge branch 'main' into deprecate-extensions

jakelishman commented on qiskit/__init__.py:
```diff
@@ -66,9 +66,6 @@
 # user config
 from qiskit import user_config as _user_config
 
-# The qiskit.extensions.x imports needs to be placed here due to the
-# mechanism for adding gates dynamically.
-import qiskit.extensions
```
Removing this in conjunction with all the rest of the work on removing uses of `qiskit.extensions` actually was a bit of a breaking change, because `qiskit.extensions` is no longer available just by having done `import qiskit`.  In other words:
```python
import qiskit

qiskit.extensions.CXGate
```
would have worked without error before, and now it'll say
```python
AttributeError: module 'qiskit' has no attribute 'extensions'
```
and require you to do `import qiskit.extensions`.

That's minor and we can probably get away with it, but it was a dedicated submodule that was previously automatically imported, so just pointing out that these removals can cause issues.

jakelishman commented on qiskit/circuit/library/generalized_gates/diagonal.py:
```diff
@@ -71,25 +76,18 @@ class Diagonal(QuantumCircuit):
     `arXiv:0406176 <https://arxiv.org/pdf/quant-ph/0406176.pdf>`_
     """
 
-    def __init__(self, diag: list[complex] | np.ndarray) -> None:
-        """Create a new Diagonal circuit.
-
+    def __init__(self, diag: Sequence[complex]) -> None:
+        r"""
```
`Sequence` isn't the correct type hint.  `ndarray` is not a `Sequence`, and `Diagonal._check_input` requires the input to be `list | ndarray` anyway.

jakelishman commented on qiskit/circuit/library/generalized_gates/diagonal.py:
```diff
@@ -71,25 +76,18 @@ class Diagonal(QuantumCircuit):
+                if the number of diagonal entries is not :math:`2^k`, where :math:`k` denotes the
+                number of qubits.
         """
-        if not isinstance(diag, (list, np.ndarray)):
-            raise CircuitError("Diagonal entries must be in a list or numpy array.")
-        num_qubits = np.log2(len(diag))
-        if num_qubits < 1 or not num_qubits.is_integer():
-            raise CircuitError("The number of diagonal entries is not a positive power of 2.")
-        if not np.allclose(np.abs(diag), 1, atol=_EPS):
-            raise CircuitError("A diagonal element does not have absolute value one.")
-
```
It might have been a bit better to make any code refactors like these in a follow-up PR, rather than squashed into a +1.5k,-1.3k changeset that notionally shouldn't have any effect on the objects.

jakelishman commented on qiskit/circuit/library/generalized_gates/mcmt.py:
```diff
@@ -19,6 +19,8 @@
 from qiskit import circuit
 from qiskit.circuit import ControlledGate, Gate, Qubit, QuantumRegister, QuantumCircuit
 from qiskit.exceptions import QiskitError
+
+# pylint: disable=cyclic-import
 from ..standard_gates import XGate, YGate, ZGate, HGate, TGate, TdgGate, SGate, SdgGate
```
This extra suppression is indicative to me that we've put a gate somewhere in the wrong place - I suspect it's `UnitaryGate`, which it might have been cleaner to give special treatment, since it's used by the base `Gate` class.  `standard_gates` shouldn't need to cyclically import `generalized_gates`.

We can re-arrange things in other PRs, though, no worries.

Cryoris replied:
This _might_ also be unnecessary -- after moving the gates there were loads of cyclic imports, and I might have added one too many ignores.... 🙈

jakelishman replied:
I think it's probably a true one - `UnitaryGate` being inside `generalized_gates` is one of the things we maybe want to reconsider before the actual release; it's kind of a base gate, and while I can see the argument for calling it "generalized", it's fundamental to a lot of how synthesis and the `Gate` class works, whereas everything else in `generalized_gates` is rather higher level and should be 100% safe to import _after_ `standard_gates`.

jakelishman commented on qiskit/circuit/library/generalized_gates/unitary.py:
```diff
@@ -79,12 +85,12 @@ def __init__(self, data, label=None):
         data = numpy.array(data, dtype=complex)
         # Check input is unitary
         if not is_unitary_matrix(data):
-            raise ExtensionError("Input matrix is not unitary.")
+            raise ValueError("Input matrix is not unitary.")
```
This feels like an error people could conceivably have been catching as part of a validation workflow.  During the deprecation period, we might have wanted to use a custom error type that derived both `ExtensionError` and `ValueError`, and only switched the documentation over to saying it raised `ValueError`.

Cryoris replied:
Yeah, I pointed this out in https://github.com/Qiskit/qiskit/pull/10725#discussion_r1308772310 as I wasn't sure whether we should take this precaution 🤔

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4120,6 +4122,669 @@ def pauli(
+                else:
+                    qubits.append(tuple_element)
+
+        # catch deprecation warning from instantiating the Snapshot instruction,
+        # as a warning is already triggered from this method
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=DeprecationWarning)
+            snap = Snapshot(
+                label, snapshot_type=snapshot_type, num_qubits=len(qubits), params=params
+            )
```
We shouldn't really do this in general; it just slows things down for no real gain.  The `stacklevel` of the `Snapshot` warning should be set correctly such that the `QuantumCircuit.snapshot` method is "blamed", which will mean that the inner warning is not shown to users.

Cryoris replied:
In that case it would show the deprecations warnings of both `QC.snapshot` and `Snapshot` or only the first?

jakelishman replied:
The standard Python warning filters only show `DeprecationWarning`s if the `stacklevel` causes the "blamed" code to be in the module `__main__` (the name for the script file being executed / code typed into the interpreter / Jupyter cells / etc).  If the stack level of `Snapshot`'s warning is set correctly, then the warning will blame the `QuantumCircuit.snapshot` method, whose module is `qiskit.circuit.quantumcircuit`, and so it won't be shown to users.

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -4880,6 +5545,10 @@ def qubit_stop_time(self, *qubits: Union[Qubit, int]) -> float:
         return 0  # If there are no instructions over bits
 
 
+# isometry is an alias for iso
+QuantumCircuit.isometry = QuantumCircuit.iso
```
Not important at all, just saying: it might have been a shade cleaner to do this within the definition of `QuantumCircuit` -
```python
class QuantumCircuit:
    def iso(self, ...): ...

    # isometry is an alias for iso
    isometry = iso
```

jakelishman commented on qiskit/synthesis/discrete_basis/gate_sequence.py:
```diff
@@ -19,7 +19,7 @@
 
 from qiskit.circuit import Gate, QuantumCircuit, Qubit
-from qiskit.extensions import UnitaryGate
+from qiskit.circuit.library.generalized_gates.unitary import UnitaryGate
```
Not too important, but imo this is is too specific an import - the documented location is `qiskit.circuit.library`, and hyperspecific import locations like this are part of the reason it's difficult to reorganise code in the future.  If it doesn't import without the hyperspecificity, it's a strong sign that the objects have been put in the wrong place, and there are true cyclic-import problems.

Cryoris replied:
This might again be a side effect of Pylint's cyclic import complaints...

jakelishman replied:
Yeah, I'm saying that in this case, this would be pylint flagging something that's actually a real problem with our logical organisation - `generalized_gates` should depend on `standard_gates`, not vice versa.  `UnitaryGate` is a core gate; it's used in the definition of `Gate` itself, so it probably would have been better for it to be given special treatment in the hierarchy, such as being in the top level of `qiskit.circuit.library` or the like.

jakelishman commented on releasenotes/notes/deprecate-extensions-d7ab3fca19962e2e.yaml:
```diff
@@ -0,0 +1,40 @@
+---
+deprecations:
+  - |
+    To streamline the structure of Qiskit's gates and operations, the :mod:`qiskit.extensions`
+    module is pending deprecation and will be deprecated in a future release.
+    The following objects have been moved to :mod:`qiskit.circuit.library`
```
All of `qiskit.circuit.library` is accessible from `qiskit.extensions` as well via a star import - it might have been good to mention that _all_ not-immediately-deprecated gates from `qiskit.extensions` should be imported from `qiskit.circuit.library`.

jakelishman reviewed (commented):
Thanks Sasha, Shelly and Elena for reviewing this and Julien for doing it!  In general, this was a complex deprecation, and it looks like it's gone fairly well.

I think I was meant to look over this before it merged, just as a final check, so I've done so now, and I just left a couple of notes (nothing major that really needs changing).  Top level stuff:

- There's a lot of unrelated code changes in here (type hints, renaming standard imports, refactoring the moved gates) that probably would have been better in separate PRs, especially because some of them will have (minor) behavioural changes.
- The title of this PR / commit as it merged was kind of wrong - most of `qiskit.extensions` _isn't_ deprecated this release, it's pending because this PR is more about moving everything into new places.
